docling_core/types/experimental/document.py

```diff
@@ -0,0 +1,1106 @@
+                    if (
+                        col.bbox is not None
+                        and add_cell_location
+                        and add_page_index
+                        and len(self.prov) > 0
+                    ):
+                        page_w, page_h = doc.pages[self.prov[0].page_no].size.as_tuple()
+                        cell_loc = DocumentToken.get_location(
+                            bbox=col.bbox.to_bottom_left_origin(page_h).as_tuple(),
+                            page_w=page_w,
+                            page_h=page_h,
+                            xsize=xsize,
+                            ysize=ysize,
+                            page_i=self.prov[0].page_no,
+                        )
+                    elif (
+                        col.bbox is not None
+                        and add_cell_location
+                        and not add_page_index
+                        and len(self.prov) > 0
+                    ):
+                        page_w, page_h = doc.pages[self.prov[0].page_no].size.as_tuple()
+
+                        cell_loc = DocumentToken.get_location(
+                            bbox=col.bbox.to_bottom_left_origin(page_h).as_tuple(),
+                            page_w=page_w,
+                            page_h=page_h,
+                            xsize=xsize,
+                            ysize=ysize,
+                            page_i=-1,
+                        )
+
+                    cell_label = ""
+                    if add_cell_label:
+                        if col.column_header:
+                            cell_label = "<col_header>"
+                        elif col.row_header:
+                            cell_label = "<row_header>"
+                        elif col.row_section:
+                            cell_label = "<row_section>"
+                        else:
+                            cell_label = "<body>"
+
+                    body += f"<col_{j}>{cell_loc}{cell_label}{text}</col_{j}>"
+
+                body += f"</row_{i}>{new_line}"
+
+        body += f"{DocumentToken.END_TABLE.value}{new_line}"
+
+        return body
+
+
+class KeyValueItem(DocItem):
+    """KeyValueItem."""
+
+
+ContentItem = Union[TextItem, FigureItem, TableItem, KeyValueItem]
+
+
+class DocumentTrees(BaseModel):
+    """DocumentTrees."""
+
+    furniture: GroupItem = GroupItem(
+        name="_root_", dloc="#/furniture"
+    )  # List[RefItem] = []
+    body: GroupItem = GroupItem(name="_root_", dloc="#/body")  # List[RefItem] = []
+
+
+class PageItem(BaseModel):
+    """PageItem."""
+
+    # A page carries separate root items for furniture and body,
+    # only referencing items on the page
+    hash: str  # page hash
+    size: Size
+    image: Optional[ImageRef] = None
+    page_no: int
+
+
+class DoclingDocument(DocumentTrees):
+    """DoclingDocument."""
+
+    version: str = "0.0.1"  # = SemanticVersion(version="0.0.1")
+    description: Any
+    file_info: FileInfo
```

Comment:
we are trying to stay away from optional (=nullable) fields, to avoid users having to always check `if doc.file_info is not None`. if possible, good defaults (with the proper type) should be preferred.

> A document doesn't need to be backed by a file.
this is actually a very valid observation.

Reply:
A document does not have to be backed by a file, but it will always need a unique identifier (`document_hash`) and it should have a human-recognizable name (`filename`). Even it if is constructed out of nothing on the fly. We have the `file_info` and `description` objects to ensure we don't scatter many fields on the root between what is supposed to be content.

Reply:
> but it will always need a unique identifier (document_hash)

Until now we have had the document_hash from PDF conversion (which BTW was actually based on the original PDF bytes, not the resulting representation after conversion).

Now that we broadening Document creation beyond PDF conversion, we should consider (additionally?) having a generic identification / deduplication mechanism.

E.g. it could be provided by the Document class, as a checksum on its nodes/content (possibly incl. metadata too).
Implications:
- if it is a field, all add* methods would be responsible for recalculating/updating it
- if it is a method, there is no such need, but then doc serialization/deserialization gets bit more complicated.

Reply:
It could be a pydantic `@computed_field` instead.